src/Datadog.Trace.DuckTyping/DuckType.Methods.cs

```diff
@@ -0,0 +1,583 @@
+                        ILHelpers.WriteLoadLocal(outOrRefParameter.LocalIndex, il);
+
+                        // If we detect duck chaining we create a new proxy instance with the output of the original target method
+                        if (NeedsDuckChaining(localType, proxyArgumentType))
+                        {
+                            // We call DuckType.CreateCache<>.Create()
+                            MethodInfo getProxyMethodInfo = typeof(CreateCache<>)
+                                .MakeGenericType(proxyArgumentType).GetMethod("Create");
+
+                            il.Emit(OpCodes.Call, getProxyMethodInfo);
+                        }
+                        else
+                        {
+                            ILHelpers.TypeConversion(il, localType, proxyArgumentType);
+                        }
+
+                        // We store the value
+                        il.Emit(OpCodes.Stind_Ref);
+                    }
+                }
+
+                // Check if the target method returns something
+                if (targetMethod.ReturnType != typeof(void))
+                {
+                    // Handle the return value
+                    // Check if the type can be converted or if we need to enable duck chaining
+                    if (NeedsDuckChaining(targetMethod.ReturnType, proxyMethodDefinition.ReturnType))
+                    {
+                        // We call DuckType.CreateCache<>.Create()
+                        MethodInfo getProxyMethodInfo = typeof(CreateCache<>)
+                            .MakeGenericType(proxyMethodDefinition.ReturnType).GetMethod("Create");
+
+                        il.Emit(OpCodes.Call, getProxyMethodInfo);
+                    }
+                    else if (returnType != proxyMethodDefinition.ReturnType)
+                    {
+                        // If the type is not the expected type we try a conversion.
+                        ILHelpers.TypeConversion(il, returnType, proxyMethodDefinition.ReturnType);
+                    }
+                }
+
+                il.Emit(OpCodes.Ret);
+            }
+        }
+
+        private static MethodInfo SelectTargetMethod(Type targetType, MethodInfo proxyMethod, ParameterInfo[] proxyMethodParameters, Type[] proxyMethodParametersTypes)
+        {
+            DuckAttribute proxyMethodDuckAttribute = proxyMethod.GetCustomAttribute<DuckAttribute>(true) ?? new DuckAttribute();
+            proxyMethodDuckAttribute.Name ??= proxyMethod.Name;
+
+            MethodInfo targetMethod = null;
+
+            // Check if the duck attribute has the parameter type names to use for selecting the target method, in case of not found an exception is thrown.
+            if (proxyMethodDuckAttribute.ParameterTypeNames != null)
+            {
+                Type[] parameterTypes = proxyMethodDuckAttribute.ParameterTypeNames.Select(pName => Type.GetType(pName, true)).ToArray();
```

Comment:
`Type.GetType(pName, true)`

Isn't it a bit weird to let a runtime exception escape here?

I don't mind, but given the trouble you went through to have your own ducktyping exceptions everywhere, it surprises me 😛

Reply:
I let that to escape there but by returning null in those cases the target method is not found and then I throw the custom exception:

https://github.com/DataDog/dd-trace-dotnet/blob/97709faa337b5cfe8cc5f143de73bc4ab79410c3/src/Datadog.Trace.DuckTyping/DuckType.Methods.cs#L442-L447